pkg/local/diff.go

```diff
@@ -188,6 +186,59 @@ func Undo(c Changes) Changes {
 	return reversed
 }
 
+// WalkS3 - walk like an Egyptian... ¯\_(ツ)¯\_
+// This walker function simulates the way object listing is performed by S3. Contrary to how a standard FS walk function behaves, S3
+// does not take into consideration the directory hierarchy. Instead, object paths include the entire path relative to the root and as a result
+// the directory or "path separator" is also taken into account when providing the listing in a lexicographical order.
+func WalkS3(root string, callbackFunc func(p string, info fs.FileInfo, err error) error) error {
+	var dirQueue []string
+	err := filepath.Walk(root, func(p string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if p == root {
+			return nil
+		}
+		if info.IsDir() {
+			// TODO: We don't return dir results for the listing, how will this effect directory markers, and can we even support directory markers?
+			// Save encountered directories in a queue and compare them with the first appearance of an file in that level
+			dirQueue = append(dirQueue, p+path.Separator) // add path separator to dir name and sort it later
+			return filepath.SkipDir
+		}
+
+		// Sort the queue since adding the suffix might the order might change after adding the suffix
+		sort.Strings(dirQueue)
```

Comment:
I don't understand: we sort the queue on every object?!  Given that sorting requires at least seeing every object on the queue, this will be slow!

### Example

Consider this directory structure:
```
a0000/a
a0001/a
a0002/a
[...]
a9999/a
z0000
z0001
z0002
[...]
z9999
```
It contains 10000 directories and 20000 objects in total.  I would expect the code to try to sort all 10000 `a` directories for each of the 10000 `z` files.  Given that sorting is $\Omega(n)$ -- at the very least it needs to go over the entire queue to determine that it is already sorted -- this will perform 100M operations.

This sort of thing _can_ and _will_ happen for many cases.  For instance consider a lakeFS repository that contains two tables, each partitioned into 10000 parts.  100M operations guaranteed.

### Possible fix

Use at least a priority queue to handle these.

Reply:
We are sorting all directories that precede a given file (if exists) when walking.
This is necessary since we need to process these directories prior to processing the file itself.
Nevertheless, I have moved the sort to inside the the if condition and improved the condition itself